### ISSUE TITLE
Implement `generateAddress` API handler

### ISSUE THREAD

**Issue by aesedepece** [witnet/sheikah#391]
#### Motivation
As mentioned in #314, generation of new P2PKH addresses requires implementing a new `generateAddress` API handler.

This method will consume @kronolynx crypto library to generate new "final keys", calculate their addresses and return those to the client along the used key path.

This method does not allow choosing the entire key path but only the account index. It will check instead what's the index of the last final key and create the next one.

#### Tasks
_Someone tackling this issue shall:_
1. Read and understand our documentation on [key derivation][hdwallets], [address formatting][addresses] and [slip32].
2. Implement the `generateAddress` API handler. This will:
2.1. Parse `params` as `GenerateAddressParams`. Throw `WRONG_PARAMS` otherwise.
2.2. Check if there's an unlocked wallet. Throw `NO_UNLOCKED_WALLET` otherwise.
2.3. Check if the account index does exist. Throw `WRONG_ACCOUNT` otherwise.
2.4. Get the maximum final key index as `k` (`0` by default).
2.5. If `label` is not set or length is `0`, set it to `Payment request #${k+1}`.
2.6. Check that `amount`, if set, is a positive number (`>= 0`). Throw `NEGATIVE_AMOUNT` otherwise. _In the future, we'll also check for enough balance_.
2.7. Check that `expires`, if set, falls in the range `[Date.now(), Date.now() + 31_536_000_000]`. Throw `PAST_EXPIRATION_DATE` or `TOO_FAR_EXPIRATION_DATE` accordingly.
2.8. Create a new final key using keypath `m/3'/4919'/i'/0/k`, where `i` is the account index and `k` is the final key index.
2.9. Store the final key in app state. Throw `WALLET_UPDATE_FAILURE` otherwise.
2.10. Persist app state (write a copy of the wallet into wallet storage). Throw `WALLET_STORE_FAILURE` otherwise.
2.11. Compute final key hash, trim to 20 bytes and encode using [slip32]. Throw `ADDRESS_ENCODING_FAILURE` otherwise.
2.12. Return a `GenerateAddressSuccess` with the used KeyPath, address and creation date.
3. Add a backend route for the `generateAddress` method in [routes.ts].
4. Create pertinent tests for the handler.

#### Deliverables
_A potential PR closing this issue shall provide:_
- A `generateAddress` handler that takes `GenerateAddressParams` as parameter and returns a `GenerateAddressSuccess` object.
- A route for the `generateAddress` method in [routes.ts].
- A test battery for the `generateAddress` handler.

#### Type definitions
```ts
type GenerateAddressParams = {
  account: number // we can optionally refine this type to be a inside range [0, 4294967295],
  label?: string,
  amount?: number,
  expirationDate?: number
}

type GenerateAddressSuccess = {
  kind: "SUCCESS",
  keyPath: KeyPath,
  address: string,
  creationDate: number
}
```

[routes.ts]: https://github.com/witnet/sheikah/blob/master/app/main/api/routes.ts
[hdwallets]: https://github.com/aesedepece/WIPs/blob/wip-adansdpc-hdwallets/wip-adansdpc-hdwallets.md
[addresses]: https://github.com/aesedepece/WIPs/blob/wip-adansdpc-addresses/wip-adansdpc-addresses.md
[slip32]: https://github.com/witnet/ts-slip32

**Comment by aesedepece** [comment]
Dependency of #320 on this issue can be heavily lifted by first committing and merging type definitions.

**Comment by aesedepece** [comment]
Please note that this issue may be blocked by another issue extending the shape of the wallet so that final keys can be tagged with `label`, `amount` and `expires` metadata. Ping me if you have been assigned this issue but this comment has not been updated to reference the corresponding issue.

**Comment by anler** [comment]
Couple of comments about this:

1. `WRONG_PARAMS` is not really needed, that's part of the protocol, if the params do not validate and Invalid Params response is sent back
2. Is the `amount` a positive (amount >= 0) or a natural (amount > 0) number? Does it make sense to generate an address with `amount = 0`?
3. `NEGATIVE_AMOUNT` is also not needed for the same reason as `1.`. We can address this at the protocol level with a refinement type
4. `PAST_EXPIRATION_DATE`/`TOO_FAR_EXPIRATION_DATE` are also not needed for the same reason as `1.`. We can address this at the protocol level with a refinement type

**Comment by anler** [comment]
Correction: `NEGATIVE_AMOUNT`/`PAST_EXPIRATION_DATE` and `TOO_FAR_EXPIRATION_DATE` cannot be handled at the protocol level because these are logical errors that the user can introduce when filling the form, then, what we need is a validation step that will receive the already parsed params `GenerateAddressParams` and validate their values accumulating the errors, then return a response of say `VALIDATION_ERROR` with the list of all errors.

**Comment by mariocao** [comment]
We discussed to modify the check of the requested amount to be greater than 0, i.e. the error would be NON_POSITIVE_AMOUNT. :)
